File: backend/src/main/java/com/project_2/backend/services/UserService.java
package com.project_2.backend.services;

import com.project_2.backend.models.UserModel;
import com.project_2.backend.models.UsersEventsModel;
import com.project_2.backend.repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;


@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    UsersEventsService usersEventsService;

    @Autowired
    SignInService signInService;

    public String createUser(UserModel user) {
        user.setPasswordHash(signInService.hashPassword(user.getPasswordHash()));
        userRepository.save(user);

        return signInService.generateToken(user.getEmail());
    }

    public List<UserModel> getAllUsers() {
        return userRepository.findAll();
    }

    public UserModel getUserById(String id) {
        return userRepository.findById(id).orElse(null);
    }

    public UserModel getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    public List<UsersEventsModel> getAllUserEventIds(String id){

        return usersEventsService.getUserEventIDs(id);
    }

    public Boolean createUserEvent(String userId, String eventId){

        if (usersEventsService.addUserEvent(userId, eventId)){
            return true;
        }else{
            return false;
        }

    }

    public String signIn(String email, String password) {
        try{
            String passwordHash = signInService.hashPassword(password);

            UserModel user  = getUserByEmail(email);

            if(user!= null && user.getPasswordHash().equals(passwordHash)) {
                return signInService.generateToken(email);
            }

        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        return null;
    }


    public UserModel updateUser(UserModel user) {
        return userRepository.save(user);

    }

}

File: backend/services/userService.js
const User = require('../models/userModel');
const UserEvents = require('../models/usersEventsModel');
const SigninService = require('./signInService');

const createUser = async (userData) => {
    const hashPassword = SigninService.hashPassword(userData.passwordHash);

    const user = new User({name: userData.name, email: userData.email, passwordHash: hashPassword, isAdmin: userData.admin});

    user.save().catch(err => {
        console.log(err);
        return null;
      })

    return await SigninService.generateToken(userData.email);
    
};

const getAllUsers = async () => {
    return await User.find({}).catch(err => {
        return null;
    });
};

const getUserById = async (userId) => {
    return await User.find({ _id: userId }).catch(err => {
        console.log(err);
        return null;
    });
};

const getUserByEmail = async (email) => {
    return await User.find({ email: email}).catch(err => {
        console.log(err);
        return null;
    });
};

const getUserEvents = async (userId) => {
    return await UserEvents.find({userId: userId}).then().catch(err => {
        return null;
    });
};

const updateUser = async (userId,userData) => {
    await User.findOneAndUpdate({ _id: userId }, {name: userData.name}).catch(err => {
        console.log(err);
        return null;
      
    });

    return true;
};


const addUserEvent = async (userId, eventId) => {
    const userEvent = new UserEvents({userId: userId, eventId: eventId});

    userEvent.save().then().catch(err => {return false});

    return await true;
};

module.exports = {
    createUser,
    getAllUsers,
    getUserById,
    getUserEvents,
    addUserEvent,
    getUserByEmail,
    updateUser
}
